includes/data-port/class-sensei-data-port-job

```diff
@@ -0,0 +1,347 @@
+
+	/**
+	 * Estimate of completion percentage.
+	 *
+	 * @var float
+	 */
+	private $percentage;
+
+	/**
+	 * Sensei_Data_Port_Job constructor.
+	 *
+	 * @param string $job_id   Unique job id.
+	 * @param array $sub_jobs  A list of subjobs for this job.
+	 * @param string $json     A json string to restore internal state from.
+	 */
+	protected function __construct( $job_id, $sub_jobs, $json = '' ) {
+		$this->job_id = $job_id;
+		$this->has_changed = false;
+
+		if ( '' !== $json ) {
+			$this->restore_from_json( $json );
+		} else {
+			$this->logs           = [];
+			$this->results        = [];
+			$this->is_completed   = false;
+			$this->has_changed    = true;
+			$this->sub_jobs_state = [];
+			$this->percentage     = 0;
+
+			foreach ( [ 'questions', 'lessons', 'courses' ] as $sub_job ) {
+				if ( in_array( $sub_job, $sub_jobs, true ) ) {
+					$this->sub_jobs_state[ $sub_job ][ 'status' ] = 'running';
+				}
+			}
+		}
+
+		add_action( 'shutdown', [ $this, 'persist' ] );
+	}
+
+	/**
+	 * Restore a stored job. Returns null if the job does not exist.
+	 *
+	 * @param string $job_id  The job id.
+	 *
+	 * @return Sensei_Data_Port_Job|null instance.
+	 */
+	public static function get( $job_id ) {
+		$json = get_option( self::get_option_name(), '' );
+
+		if ( empty( $json ) ) {
+			return null;
+		}
+
+		return new static( $job_id, $json );
+	}
+
+	/**
+	 * Get the results of the job.
+	 *
+	 * @return array The results.
+	 */
+	public function get_results() {
+		return $this->results;
+	}
+
+	/**
+	 * Get the logs of the job.
+	 *
+	 * @param int $offset  Offset to support pagination.
+	 * @param int $limit   Limit to support pagination.
+	 *
+	 * @return array The logs.
+	 */
+	public function get_logs( $offset = 0 , $limit = 0 ) {
+		if ( $limit > 0 ) {
+			return array_slice( $this->logs, $offset, $limit );
+		} else {
+			return $this->logs;
+		}
+	}
+
+	/**
+	 * Delete any stored state for this job.
+	 */
+	public function clean_up() {
+		delete_option( self::get_option_name( $this->job_id ) );
+	}
+
+	/**
+	 * Get the completion status of the job.
+	 *
+	 * @return array
+	 */
+	public function get_status() {
+		if ( $this->is_completed ) {
+			return [
+				'status' => 'completed',
+				'percentage' => 100,
+			];
+		}
+
+		return [
+			'status' => 'running',
+			'percentage' => $this->percentage,
+		];
+	}
+
+	/**
+	 * Creates a new data port job.
+	 *
+	 * @param string $job_id    Unique id of the job.
+	 * @param array $sub_jobs   Sub-job levels.
+	 *
+	 * @return Sensei_Data_Port_Job The newly created job.
+	 */
+	abstract public static function create( $job_id, $sub_jobs );
+
+	/**
+	 * Do question porting.
+	 *
+	 * @param array $question_state  Question sub-job state to be maintained in subclasses. Should set 'status' to
+	 *                               'completed' when the port is completed.
+	 *
+	 * @return float Percentage that has been done.
+	 */
+	abstract protected function do_question_port( &$question_state );
```

Comment:
These could probably be defined as the callbacks to the tasks. Import/export may have different tasks. For example, export will have the additional task of packaging.

Reply:
Can you give an example of the callback solution? We could possibly add a post process step that would allow packaging.